rfcs/0097-unify-packed-and-object.md

```diff
@@ -0,0 +1,677 @@
+
+There are several design choices in terms of C API convention in light of this new proposal. They will affect the internal data layout of TVMArgs.
+
+- **D4a:** Current C API:
+    
+    ```cpp
+    int TVMCPackedFunc(PackedFuncHandle handle,
+                       int num_args, int* type_codes, TVMValue* values,
+                       int* out_tcode, TVMValue* out_value);
+    ```
+    
+    - This would require packing unpacking the typecode. Note that small string passing wont work because of the lack of the seq_len padding field.
+- **D4b:** Combine type code and TVMValue
+    
+    ```cpp
+    
+    // TVMAnyView and TVMAnyValue follows the same layout as AnyPodBaseValue_
+    int TVMCPackedFunc(PackedFuncHandle handle,
+                       int num_args, TVMAnyView* args, 
+                       TVMAnyValue* out);
+    ```
+    
+    This approach combines code and value into a single entity, this would mean a change of ABI convention in generated code. This approach makes it possible to directly return a small-str without boxing it (however if it is faced in python frontend, likely we still need to box it to str).
+    
+    We will introduce adapters to support the old calling convention, which constructs TVMAnyView on the stack then pass things over in the transitioning period.
+    
+
+## Prior Arts
+
+This RFC is a further evolutions of TVM’s Packed and Object System. We also learn lessons from related projects such as matxscript, which demonstrates real world use-case motivations for some of the design perspectives
+
+## Rationales and Alternatives
+
+There are several design choices we listed in the above design points, we summarize them here, provide our rationales and recommendations.
+
+### D0a and D0b: Default size of AnyValue
+
+The default size of AnyValue
+
+- **D0a**: 16 bytes
+    - AnyView also cost 16bytes. Use AnyPad<n> in locals when necessary
+- **D0b**: 24 bytes
+
+Both choices are likely OK and won’ have big impact. Because we support AnyPad<n> natively in locals (and can also pass AnyPad<n> around) by taking address of it, we would recommend start with D0a — this also keeps the overall call stack cost consistent with the current design.
+
+### D1a and D1b: Any.type_code for Object
+
+When to store object type code in AnyValue.type_code. Type code segmentation organization
+
+- **D1a**: Stores type_code if the object is in S1
+- **D1b**: Stores all type_code (S1-S4)
+- **D1c**: Store type of only static segment of types. (S1-S3 but not S4).
+
+The current state starts with D1a.  The tradeoff here is again not as critical. One thing to consider is the object code lookup cost when the frontend only recognizes part of the type. Likely starting from D1a is a reasonable pt.
+
+### D1section: Any.type_code section convention
+
+When to store non-object type code in AnyValue.type_code. We hae the following code segmentation organization 
```

Comment:
```suggestion
When to store non-object type code in AnyValue.type_code. We have the following code segmentation organization 
```

Reply:
Thanks for the suggestions, updated accoridngly